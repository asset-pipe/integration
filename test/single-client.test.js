'use strict';

const { resolve } = require;
const { spawn } = require('child_process');
const Client = require('@asset-pipe/client');
const buildServerUri = 'http://127.0.0.1:7100';
const client = new Client({ buildServerUri });
const supertest = require('supertest');
const request = supertest(buildServerUri);

const localiseBodyPaths = body =>
    JSON.parse(
        JSON.stringify(body, null, 2).replace(
            /"file":\s".*\/asset-pipe\//g,
            '"file": "'
        )
    );

let server;
beforeAll(() => {
    jest.setTimeout(20000);
    return new Promise((resolve, reject) => {
        server = spawn('./node_modules/.bin/asset-pipe-server', [], {
            env: Object.assign({}, process.env, {
                NODE_ENV: 'development',
            }),
        });
        server.stdout.once('data', () => resolve());
        server.stderr.once('data', () => resolve());
        server.once('error', err => reject(err));
        server.once('close', () => resolve());
    });
});
afterAll(() => {
    server.kill();
});

test('Client uploads a js feed to build server', async () => {
    expect.assertions(1);
    const result = await client.uploadFeed([resolve('../assets/index.js')]);
    expect(result).toMatchSnapshot();
});

test('Client gets uploaded js feed from build server', async () => {
    expect.assertions(1);
    const result = await client.uploadFeed([resolve('../assets/index.js')]);
    const { body } = await request
        .get(`/feed/${result.file}`)
        .set('Accept', 'application/json')
        .expect('Content-Type', /application\/json/)
        .expect(200);
    expect(localiseBodyPaths(body)).toMatchSnapshot();
});

test('Client uploads a css feed to build server', async () => {
    expect.assertions(1);
    const result = await client.uploadFeed([resolve('../assets/style.css')]);
    expect(result).toMatchSnapshot();
});

test('Client gets uploaded css feed from build server', async () => {
    expect.assertions(1);
    const result = await client.uploadFeed([resolve('../assets/style.css')]);
    const { body } = await request
        .get(`/feed/${result.file}`)
        .set('Accept', 'application/json')
        .expect('Content-Type', /application\/json/)
        .expect(200);
    expect(body).toMatchSnapshot();
});

test('Client requests bundling of a js feed on build server', async () => {
    expect.assertions(1);
    const uploadResponse = await client.uploadFeed([
        resolve('../assets/index.js'),
    ]);
    const bundleResponse = await client.createRemoteBundle(
        [uploadResponse.file],
        'js'
    );
    expect(bundleResponse).toMatchSnapshot();
});

test('Client gets bundled js file from build server', async () => {
    expect.assertions(1);
    const uploadResponse = await client.uploadFeed([
        resolve('../assets/index.js'),
    ]);
    const bundleResponse = await client.createRemoteBundle(
        [uploadResponse.file],
        'js'
    );
    const { text } = await request
        .get(`/bundle/${bundleResponse.file}`)
        .expect(200);
    expect(text).toMatchSnapshot();
});

test('Client requests bundling of a css feed on build server', async () => {
    expect.assertions(1);
    const uploadResponse = await client.uploadFeed([
        resolve('../assets/style.css'),
    ]);
    const bundleResponse = await client.createRemoteBundle(
        [uploadResponse.file],
        'css'
    );
    expect(bundleResponse).toMatchSnapshot();
});

test('Client gets bundled css file from build server', async () => {
    expect.assertions(1);
    const uploadResponse = await client.uploadFeed([
        resolve('../assets/style.css'),
    ]);
    const bundleResponse = await client.createRemoteBundle(
        [uploadResponse.file],
        'css'
    );
    const { text } = await request
        .get(`/bundle/${bundleResponse.file}`)
        .expect(200);
    expect(text).toMatchSnapshot();
});

test('Uploaded js feed from build server is not deduped', async () => {
    expect.assertions(1);
    const result = await client.uploadFeed([
        resolve('../assets/duped-before-uploading.js'),
    ]);
    const { body } = await request
        .get(`/feed/${result.file}`)
        .set('Accept', 'application/json')
        .expect('Content-Type', /application\/json/)
        .expect(200);
    expect(localiseBodyPaths(body)).toMatchSnapshot();
});

test('Bundled js feed from build server is deduped', async () => {
    expect.assertions(1);
    const uploadResponse = await client.uploadFeed([
        resolve('../assets/duped-before-uploading.js'),
    ]);
    const bundleResponse = await client.createRemoteBundle(
        [uploadResponse.file],
        'js'
    );
    const { text } = await request
        .get(`/bundle/${bundleResponse.file}`)
        .expect(200);
    expect(text).toMatchSnapshot();
});

test('NODE_ENV variables replaced during bundling', async () => {
    expect.assertions(1);
    const uploadResponse = await client.uploadFeed([
        resolve('../assets/node_envs.js'),
    ]);
    const bundleResponse = await client.createRemoteBundle(
        [uploadResponse.file],
        'js'
    );
    const { text } = await request
        .get(`/bundle/${bundleResponse.file}`)
        .expect(200);
    expect(text).not.toMatch(`process.env.NODE_ENV === 'production'`);
});
