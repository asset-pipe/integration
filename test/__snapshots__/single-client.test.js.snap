// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bundled js feed from build server is deduped 1`] = `
"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(1, 2);

},{\\"./d\\":2}],2:[function(require,module,exports){
'use strict';

module.exports = function sum(a, b) {
    return a + b;
};

},{}],3:[function(require,module,exports){
require('./a');

},{\\"./a\\":1}],4:[function(require,module,exports){
arguments[4][3][0].apply(exports,arguments)
},{\\"./a\\":1,\\"dup\\":3}],5:[function(require,module,exports){
arguments[4][3][0].apply(exports,arguments)
},{\\"./a\\":1,\\"dup\\":3}],6:[function(require,module,exports){
const a = require('./dup-1');
const b = require('./dup-2');
const c = require('./dup-3');

},{\\"./dup-1\\":3,\\"./dup-2\\":4,\\"./dup-3\\":5}]},{},[6]);
"
`;

exports[`Client gets bundled css file from build server 1`] = `
"html, body, div, span, object, iframe,
h1, h2, h3, h4, h5, h6, p, blockquote, pre,
abbr, address, cite, code,
del, dfn, em, img, ins, kbd, q, samp,
small, strong, sub, sup, var,
b, i,
dl, dt, dd, ol, ul, li,
fieldset, form, label, legend,
table, caption, tbody, tfoot, thead, tr, th, td,
article, aside, canvas, details, figcaption, figure, 
footer, header, hgroup, menu, nav, section, summary,
time, mark, audio, video {
    margin:0;
    padding:0;
    border:0;
    outline:0;
    font-size:100%;
    vertical-align:baseline;
    background:transparent;
}

body {
    line-height:1;
}

article,aside,details,figcaption,figure,
footer,header,hgroup,menu,nav,section { 
    display:block;
}

nav ul {
    list-style:none;
}

blockquote, q {
    quotes:none;
}

blockquote:before, blockquote:after,
q:before, q:after {
    content:'';
    content:none;
}

a {
    margin:0;
    padding:0;
    font-size:100%;
    vertical-align:baseline;
    background:transparent;
}

/* change colours to suit your needs */

ins {
    background-color:#ff9;
    color:#000;
    text-decoration:none;
}

/* change colours to suit your needs */

mark {
    background-color:#ff9;
    color:#000; 
    font-style:italic;
    font-weight:bold;
}

del {
    text-decoration: line-through;
}

abbr[title], dfn[title] {
    border-bottom:1px dotted;
    cursor:help;
}

table {
    border-collapse:collapse;
    border-spacing:0;
}

/* change border colour to suit your needs */

hr {
    display:block;
    height:1px;
    border:0;   
    border-top:1px solid #cccccc;
    margin:1em 0;
    padding:0;
}

input, select {
    vertical-align:middle;
}

h1 {
    color: #000000;
    size: 32px;
}"
`;

exports[`Client gets bundled js file from build server 1`] = `
"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

const { hello } = require('./lib');

console.log(hello());

},{\\"./lib\\":2}],2:[function(require,module,exports){
'use strict';

module.exports.hello = function hello() {
    console.log('hello');
};

},{}]},{},[1]);
"
`;

exports[`Client gets uploaded css feed from build server 1`] = `
Array [
  Object {
    "content": "html, body, div, span, object, iframe,
h1, h2, h3, h4, h5, h6, p, blockquote, pre,
abbr, address, cite, code,
del, dfn, em, img, ins, kbd, q, samp,
small, strong, sub, sup, var,
b, i,
dl, dt, dd, ol, ul, li,
fieldset, form, label, legend,
table, caption, tbody, tfoot, thead, tr, th, td,
article, aside, canvas, details, figcaption, figure, 
footer, header, hgroup, menu, nav, section, summary,
time, mark, audio, video {
    margin:0;
    padding:0;
    border:0;
    outline:0;
    font-size:100%;
    vertical-align:baseline;
    background:transparent;
}

body {
    line-height:1;
}

article,aside,details,figcaption,figure,
footer,header,hgroup,menu,nav,section { 
    display:block;
}

nav ul {
    list-style:none;
}

blockquote, q {
    quotes:none;
}

blockquote:before, blockquote:after,
q:before, q:after {
    content:'';
    content:none;
}

a {
    margin:0;
    padding:0;
    font-size:100%;
    vertical-align:baseline;
    background:transparent;
}

/* change colours to suit your needs */

ins {
    background-color:#ff9;
    color:#000;
    text-decoration:none;
}

/* change colours to suit your needs */

mark {
    background-color:#ff9;
    color:#000; 
    font-style:italic;
    font-weight:bold;
}

del {
    text-decoration: line-through;
}

abbr[title], dfn[title] {
    border-bottom:1px dotted;
    cursor:help;
}

table {
    border-collapse:collapse;
    border-spacing:0;
}

/* change border colour to suit your needs */

hr {
    display:block;
    height:1px;
    border:0;   
    border-top:1px solid #cccccc;
    margin:1em 0;
    padding:0;
}

input, select {
    vertical-align:middle;
}

h1 {
    color: #000000;
    size: 32px;
}",
    "file": "integration/assets/style.css",
    "id": "8b30de6afdb4474cfd5c5df052408dc6ee170d243861c3fa63d246440b5f472d",
    "name": "integration",
    "version": "1.0.0",
  },
]
`;

exports[`Client gets uploaded js feed from build server 1`] = `
Array [
  Object {
    "deps": Object {
      "./lib": "e2e8400ede1e63e031135860fa638bac1a1dd8a0bccf4f0f5b283c439672c6ea",
    },
    "entry": true,
    "expose": false,
    "file": "integration/assets/index.js",
    "id": "a394b4261a9e731e26fbd22f93ca036b458ac46521826424134bdb32344c8ef8",
    "index": 1,
    "indexDeps": Object {
      "./lib": 2,
    },
    "order": 0,
    "source": "'use strict';

const { hello } = require('./lib');

console.log(hello());
",
  },
  Object {
    "deps": Object {},
    "file": "integration/assets/lib.js",
    "id": "e2e8400ede1e63e031135860fa638bac1a1dd8a0bccf4f0f5b283c439672c6ea",
    "index": 2,
    "indexDeps": Object {},
    "source": "'use strict';

module.exports.hello = function hello() {
    console.log('hello');
};
",
  },
]
`;

exports[`Client requests bundling of a css feed on build server 1`] = `
Object {
  "file": "8b30de6afdb4474cfd5c5df052408dc6ee170d243861c3fa63d246440b5f472d.css",
  "uri": "http://127.0.0.1:7100/bundle/8b30de6afdb4474cfd5c5df052408dc6ee170d243861c3fa63d246440b5f472d.css",
}
`;

exports[`Client requests bundling of a js feed on build server 1`] = `
Object {
  "file": "8b37efa3105ca5e99dceeed61283bb3e547af6b2f5ad911eadb8b85b42885eee.js",
  "uri": "http://127.0.0.1:7100/bundle/8b37efa3105ca5e99dceeed61283bb3e547af6b2f5ad911eadb8b85b42885eee.js",
}
`;

exports[`Client uploads a css feed to build server 1`] = `
Object {
  "file": "80c8a092aee329dd8cf852eaa081829bcac6f24bfa4ef11b90c94b9607244251.json",
  "id": "80c8a092aee329dd8cf852eaa081829bcac6f24bfa4ef11b90c94b9607244251",
  "uri": "http://127.0.0.1:7100/feed/80c8a092aee329dd8cf852eaa081829bcac6f24bfa4ef11b90c94b9607244251.json",
}
`;

exports[`Client uploads a js feed to build server 1`] = `
Object {
  "file": "5c0319bc8def11e319cee258be0f2f8a970208042d6dc3073b40ef5d3a8098a6.json",
  "id": "5c0319bc8def11e319cee258be0f2f8a970208042d6dc3073b40ef5d3a8098a6",
  "uri": "http://127.0.0.1:7100/feed/5c0319bc8def11e319cee258be0f2f8a970208042d6dc3073b40ef5d3a8098a6.json",
}
`;

exports[`Uploaded js feed from build server is not deduped 1`] = `
Array [
  Object {
    "deps": Object {
      "./d": "a83a8f5a3f18d24f036db81d8660b46af30b75902500f405df28b8c368045add",
    },
    "file": "integration/assets/a.js",
    "id": "3b0887ddfac98bb9c42ae30bdc580452e2798324bd16572e4635a7098e2139be",
    "index": 1,
    "indexDeps": Object {
      "./d": 2,
    },
    "source": "'use strict';

const sum = require('./d');

sum(1, 2);
",
  },
  Object {
    "deps": Object {},
    "file": "integration/assets/d.js",
    "id": "a83a8f5a3f18d24f036db81d8660b46af30b75902500f405df28b8c368045add",
    "index": 2,
    "indexDeps": Object {},
    "source": "'use strict';

module.exports = function sum(a, b) {
    return a + b;
};
",
  },
  Object {
    "deps": Object {
      "./a": "3b0887ddfac98bb9c42ae30bdc580452e2798324bd16572e4635a7098e2139be",
    },
    "file": "integration/assets/dup-1.js",
    "id": "b584f0bd0b03f3d68d5db58c1960e4d7f2629152500aea2fb60bf686086e8f40",
    "index": 3,
    "indexDeps": Object {
      "./a": 1,
    },
    "source": "require('./a');
",
  },
  Object {
    "deps": Object {
      "./a": "3b0887ddfac98bb9c42ae30bdc580452e2798324bd16572e4635a7098e2139be",
    },
    "file": "integration/assets/dup-2.js",
    "id": "b584f0bd0b03f3d68d5db58c1960e4d7f2629152500aea2fb60bf686086e8f40",
    "index": 4,
    "indexDeps": Object {
      "./a": 1,
    },
    "source": "require('./a');
",
  },
  Object {
    "deps": Object {
      "./a": "3b0887ddfac98bb9c42ae30bdc580452e2798324bd16572e4635a7098e2139be",
    },
    "file": "integration/assets/dup-3.js",
    "id": "b584f0bd0b03f3d68d5db58c1960e4d7f2629152500aea2fb60bf686086e8f40",
    "index": 5,
    "indexDeps": Object {
      "./a": 1,
    },
    "source": "require('./a');
",
  },
  Object {
    "deps": Object {
      "./dup-1": "b584f0bd0b03f3d68d5db58c1960e4d7f2629152500aea2fb60bf686086e8f40",
      "./dup-2": "b584f0bd0b03f3d68d5db58c1960e4d7f2629152500aea2fb60bf686086e8f40",
      "./dup-3": "b584f0bd0b03f3d68d5db58c1960e4d7f2629152500aea2fb60bf686086e8f40",
    },
    "entry": true,
    "expose": false,
    "file": "integration/assets/duped-before-uploading.js",
    "id": "9046c91ca8a6fe5989543e1dbb411f52700f8ee0b140ccf0c18b3fa3855bd1c0",
    "index": 6,
    "indexDeps": Object {
      "./dup-1": 3,
      "./dup-2": 4,
      "./dup-3": 5,
    },
    "order": 0,
    "source": "const a = require('./dup-1');
const b = require('./dup-2');
const c = require('./dup-3');
",
  },
]
`;
